lib/class-wp-rest-shortcodes-controller.php

```diff
@@ -0,0 +1,189 @@
+		$shortcode             = isset( $args['shortcode'] ) ? trim( $args['shortcode'] ) : '';
+
+		// Initialize $data.
+		$data = array(
+			'html'                  => $output,
+			'type'                  => $type,
+			'head_scripts_styles'   => $head_scripts_styles,
+			'footer_scripts_styles' => $footer_scripts_styles,
+		);
+
+		if ( empty( $shortcode ) ) {
+			$data['html'] = __( 'Enter something to preview', 'gutenberg' );
+			return rest_ensure_response( $data );
+		}
+
+		if ( ! empty( $post ) ) {
+			setup_postdata( $post );
+		}
+
+		// Since the [embed] shortcode needs to be run earlier than other shortcodes.
+		if ( has_shortcode( $shortcode, 'embed' ) ) {
+			$output = $wp_embed->run_shortcode( $shortcode );
+		} else {
+			$output = do_shortcode( $shortcode );
+		}
+
+		if ( empty( $output ) ) {
+			$data['html'] = __( 'Sorry, couldn\'t render a preview', 'gutenberg' );
+			return rest_ensure_response( $data );
+		}
+
+		ob_start();
+		wp_head();
+		$head_scripts_styles = ob_get_clean();
+
+		ob_start();
+		wp_footer();
+		$footer_scripts_styles = ob_get_clean();
+
+		// Check if shortcode is returning a video. The video type will be used by the frontend to maintain 16:9 aspect ratio.
+		if ( has_shortcode( $shortcode, 'video' ) ) {
+			$type = 'video';
+		} elseif ( has_shortcode( $shortcode, 'embed' ) ) {
+			$embed_request = new WP_REST_Request( 'GET', '/oembed/1.0/proxy' );
+			$pattern       = get_shortcode_regex();
+			if ( preg_match_all( '/' . $pattern . '/s', $shortcode, $matches ) ) {
+				$embed_request['url'] = $matches[5][0];
+				$embed_response       = rest_do_request( $embed_request );
+				if ( $embed_response->is_error() ) {
+					$data['html'] = __( 'Sorry, couldn\'t render a preview', 'gutenberg' );
+					return rest_ensure_response( $data );
+				}
+				$embed_data = $embed_response->get_data();
+			}
+			$type = $embed_data->type;
+		}
+		$data = array(
+			'html'                  => $output,
+			'head_scripts_styles'   => $head_scripts_styles,
+			'footer_scripts_styles' => $footer_scripts_styles,
+			'type'                  => $type,
+		);
+
+		return rest_ensure_response( $data );
+	}
+
+	/**
+	 * Retrieves a shortcode block's schema, conforming to JSON Schema.
+	 *
+	 * @since 0.10.0
+	 * @access public
+	 *
+	 * @return array Item schema data.
+	 */
+	public function get_item_schema() {
+		return array(
+			'$schema'    => 'http://json-schema.org/schema#',
+			'title'      => 'shortcode-block',
+			'type'       => 'object',
+			'properties' => array(
+				'html'                  => array(
+					'description' => __( 'The block\'s content with shortcodes filtered through hooks.', 'gutenberg' ),
+					'type'        => 'string',
+					'required'    => true,
+				),
+				'type'                  => array(
+					'description' => __( 'The filtered content type - video or otherwise', 'gutenberg' ),
+					'type'        => 'string',
+					'required'    => true,
+				),
+				'head_scripts_styles'   => array(
```

Comment:
Technically this will contain not just scripts and styles but anything that is output by `wp_head` (and `wp_footer`) below. I think this would be better named just `html_head` and `html_footer`, and maybe `html_body` instead of `html`.